examples/async/vectors.py

```diff
@@ -0,0 +1,170 @@
+    async_connections,
+)
+
+DATASET_URL = "https://raw.githubusercontent.com/elastic/elasticsearch-labs/main/datasets/workplace-documents.json"
+MODEL_NAME = "all-MiniLM-L6-v2"
+
+# initialize sentence tokenizer
+nltk.download("punkt", quiet=True)
+
+
+class Passage(InnerDoc):
+    content = Text()
+    embedding = DenseVector()
+
+
+class WorkplaceDoc(AsyncDocument):
+    class Index:
+        name = "workplace_documents"
+
+    name = Text()
+    summary = Text()
+    content = Text()
+    created = Date()
+    updated = Date()
+    url = Keyword()
+    category = Keyword()
+    passages = Nested(Passage)
+
+    _model = None
+
+    @classmethod
+    def get_embedding_model(cls):
+        if cls._model is None:
+            cls._model = SentenceTransformer(MODEL_NAME)
+        return cls._model
+
+    def clean(self):
+        # split the content into sentences
+        passages = nltk.sent_tokenize(self.content)
+
+        # generate an embedding for each passage and save it as a nested document
+        model = self.get_embedding_model()
+        for passage in passages:
+            self.passages.append(
+                Passage(content=passage, embedding=list(model.encode(passage)))
+            )
+
+
+async def create():
+
+    # create the index
+    await WorkplaceDoc._index.delete(ignore_unavailable=True)
+    await WorkplaceDoc.init()
+
+    # download the data
+    dataset = json.loads(urlopen(DATASET_URL).read())
+
+    # import the dataset
+    for data in tqdm(dataset, desc="Indexing documents..."):
+        doc = WorkplaceDoc(
+            name=data["name"],
+            summary=data["summary"],
+            content=data["content"],
+            created=data.get("created_on"),
+            updated=data.get("updated_at"),
+            url=data["url"],
+            category=data["category"],
+        )
+        await doc.save()
+
+
+async def search(query):
+    model = WorkplaceDoc.get_embedding_model()
+    search = WorkplaceDoc.search().knn(
+        field="passages.embedding",
+        k=5,
+        num_candidates=50,
+        query_vector=list(model.encode(query)),
+        inner_hits={"size": 3},
+    )
+    async for hit in search:
+        print(f"Document: {hit.name} (Category: {hit.category}")
+        for passage in hit.meta.inner_hits.passages:
+            print(f"  - [Score: {passage.meta.score}] {passage.content!r}")
+        print("")
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description="Vector database with Elasticsearch")
+    parser.add_argument(
+        "--create", action="store_true", help="Create and populate a new index"
+    )
+    parser.add_argument("query", action="store", help="The search query")
+    return parser.parse_args()
+
+
+async def main():
+    args = parse_args()
+
+    # initiate the default connection to elasticsearch
+    async_connections.create_connection(hosts=[os.environ["ELASTICSEARCH_URL"]])
+
+    if args.create or not await WorkplaceDoc._index.exists():
```

Comment:
Not sure what the point of create is if we're going to create anyway? Is that meant to "recreate"?

Reply:
Yes. I should rename this `--recreate` for more clarity.